File: 0047-permutations-ii/0047-permutations-ii.cpp
class Solution {

private:

    void getPermute(vector<vector<int>>& result, vector<int>& nums, int start) {
        if (start == nums.size() - 1) {
            result.push_back(nums);
            return;
        }

        unordered_set<int> used;  // To track numbers that have been used at the current level

        for (int i = start; i < nums.size(); i++) {
            if (used.count(nums[i])) {
                continue;  // Skip if the number has already been used in this position
            }

            used.insert(nums[i]);  // Mark the number as used

            swap(nums[start], nums[i]);
            
            getPermute(result, nums, start + 1);
            
            swap(nums[start], nums[i]);  // Backtrack
        }
        return;
    }

public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> result;

        sort(nums.begin(), nums.end());  // Sort to handle duplicates

        getPermute(result, nums, 0);
        return result;
    }
};

File: 0047-permutations-ii/0047-permutations-ii.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        # how to prvent duplicate permutaion -> to prevent selecting the same number in the same level
        # how skip the duplicate num in currrent level

        nums.sort()

        def backtrack(start, result):
            if start == len(nums):
                result.append(nums[:])
                return

            used = set()

            for i in range(start, len(nums)):
                if nums[i] in used:
                    continue
                
                used.add(nums[i])
                nums[i], nums[start] = nums[start], nums[i]

                backtrack(start + 1, result)

                nums[i], nums[start] = nums[start], nums[i]


        result = []
        backtrack(0, result)

        return result
        